templated-this.md

```diff
@@ -1,151 +1,315 @@
-  string s = "Hello, world!";
-  
-  // Here, the closure object is an rvalue:
-  // the captured s will be correctly moved
-  bar([s] (auto&& this) { 
-    return std::forward<decltype(*this)>(*this).s;
-  }
-}
-```
-
-From these examples it follows that a definition of a member function with `T&&
-this` as its first parameter may be likened to declaring a free function
-template that is a friend of `T` and that is constrained to take a forwarding
-reference to `T` as its first argument. Usage of `*this` within the function is
-very similar to existing practice dealing with arguments that are forwarding
-references.
-
-It is also clear that this opens the ability to refer to a closure object from
-inside a lambda expression: something that until now has been impossible. We
-believe this to be important for the performance reasons given. However, this is
-perhaps undesirable in general and has the potential for poor interaction with
-by-reference captures and existing `this` or `*this` capture. For this reason we
-propose allowing using `*this` only to refer to by-copy captures inside a lambda
-expression body.
-
-We believe that existing `this` and `*this` capture is entirely obviated by init
-capture and thus need not be used; any new code using `auto&& this` to access
-the closure object has no need of `this` or `*this` capture.
-
-TBD: Do we believe this?
-
-## Design Decisions
-   
-In addition to solving the existing problems, desirable properties of a solution
-are:
-
-* It should work the same way for member functions and lambda expressions.
-
-* It should work like existing practice as much as possible, while adding as
-  little extra syntax as possible. Using `auto&& this` follows existing patterns
-  for dealing with forwarding references. Optionally adding "this" as the first
-  parameter fits with many programmers' mental model of the `this` pointer being
-  the first parameter to member functions "under the hood" and is comparable to
-  usage in other languages, e.g. Python.
-
-* It should avoid adding extraneous syntax to function declarations. In a world
-  where every function is fast becoming tagged with attributes, constexpr,
-  noexcept specifications, and (this proposal notwithstanding) reference
-  qualifiers, readable, noise-free function declarations are to be prized.
-
-TBD: more here.
-
-## Impact on the Standard
-   
-TBD: A bunch of stuff in section 8.1.5 [expr.prim.lambda].
-
-## Acknowledgments
-
-Thanks to the following for their help and guidance:
-
-Louis Dionne
-
-## References
-
-We should probably look up the paper that added `*this` capture and why. And any
-other salient papers.
-
+<pre>
+Document number: PxxxxR0
+Date:            2017-10-03
+Project:         Programming Language C++, Evolution Working Group
+Reply-to:        Ben Deane <ben at elbeno dot com>              
+                 Gašper Ažman <gasper dot azman at gmail dot com>
+</pre>
+
+Template `this`! (Divining the value-category of `*this`)
```

Comment:
:+1: I like this title.

Reply:
:) I thought you might. Need to figure out how to do subtitles.

Reply:
If you prefer we could switch to LaTeX. Michael Park has a good LaTeX setup which I copied for https://github.com/elbeno/wg21 (I'm also writing a constexpr findings paper with Jason). I'm fine with either markdown or LaTeX.

Reply:
I would, but not while we have substantive changes here. Perhaps once this is merged you could swing that?